<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Brings SpatialExperiment to the tidyverse  â€¢ tidySpatialExperiment</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Brings SpatialExperiment to the tidyverse ">
<meta property="og:description" content="tidySpatialExperiment provides a bridge between the SpatialExperiment package and the 
    tidyverse ecosystem. It creates an invisible layer that allows you to interact with a 
    SpatialExperiment object as if it were a tibble; enabling the use of functions from dplyr, 
    tidyr, ggplot2 and plotly. But, underneath, your data remains a SpatialExperiment object.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">tidySpatialExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/overview.html">Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/william-hutchison/tidySpatialExperiment/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="tidyspatialexperiment---part-of-tidyomics-">tidySpatialExperiment - part of <em>tidyomics</em> <a><img src="reference/figures/logo.png" align="right" height="138"></a>
<a class="anchor" aria-label="anchor" href="#tidyspatialexperiment---part-of-tidyomics-"></a>
</h1></div>
<!-- badges: start -->

<p>Resources to help you get started with tidySpatialExperiment and <em>tidyomics</em>:</p>
<ul>
<li><a href="http://william-hutchison.github.io/tidySpatialExperiment/" class="external-link">The tidySpatialExperiment website</a></li>
<li><a href="https://tidyomics.github.io/tidyomicsBlog/" class="external-link">The tidyomics blog</a></li>
<li><a href="https://rstudio-pubs-static.s3.amazonaws.com/792462_f948e766b15d4ee5be5c860493bda0b3.html" class="external-link">Third party tutorials</a></li>
</ul>
<p>The <em>tidyomics</em> ecosystem includes packages for:</p>
<ul>
<li>
<p>Working with genomic features:</p>
<ul>
<li>
<a href="https://github.com/sa-lee/plyranges" class="external-link">plyranges</a>, for tidy manipulation of genomic range data.</li>
<li>
<a href="https://github.com/nullranges/nullranges" class="external-link">nullranges</a>, for tidy generation of genomic ranges representing the null hypothesis.</li>
<li>
<a href="https://github.com/tidyomics/plyinteractions" class="external-link">plyinteractions</a>, for tidy manipulation of genomic interaction data.</li>
</ul>
</li>
<li>
<p>Working with transcriptomic features:</p>
<ul>
<li>
<a href="https://github.com/stemangiola/tidySummarizedExperiment" class="external-link">tidySummarizedExperiment</a>, for tidy manipulation of SummarizedExperiment objects.</li>
<li>
<a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a>, for tidy manipulation of SingleCellExperiment objects.</li>
<li>
<a href="https://github.com/stemangiola/tidyseurat" class="external-link">tidyseurat</a>, for tidy manipulation of Seurat objects.</li>
<li>
<a href="https://github.com/stemangiola/tidybulk" class="external-link">tidybulk</a>, for bulk RNA-seq analysis.</li>
</ul>
</li>
<li>
<p>Working with cytometry features:</p>
<ul>
<li>
<a href="https://github.com/keyes-timothy/tidytof" class="external-link">tidytof</a>, for tidy manipulation of high-dimensional cytometry data.</li>
</ul>
</li>
</ul>
<p>A few more tidy tools for data manipulation and plotting:</p>
<ul>
<li>
<a href="https://github.com/stemangiola/tidyHeatmap" class="external-link">tidyHeatmap</a>, for producing heatmaps with tidy principles. analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate" class="external-link">tidygate</a>, for interactive plotting and gating.</li>
</ul>
</div>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>tidySpatialExperiment provides a bridge between the <a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a> [@righelli2022spatialexperiment] package and the <a href="https://www.tidyverse.org" class="external-link"><em>tidyverse</em></a> [@wickham2019welcome] ecosystem. It creates an invisible layer that allows you to interact with a SpatialExperiment object as if it were a tibble; enabling the use of functions from <a href="https://github.com/tidyverse/dplyr" class="external-link">dplyr</a>, <a href="https://github.com/tidyverse/tidyr" class="external-link">tidyr</a>, <a href="https://github.com/tidyverse/ggplot2" class="external-link">ggplot2</a> and <a href="https://github.com/plotly/plotly.R" class="external-link">plotly</a>. But, underneath, your data remains a SpatialExperiment object.</p>
<p>tidySpatialExperiment also provides three additional utility functions.</p>
<div class="section level2">
<h2 id="functions-and-utilities">Functions and utilities<a class="anchor" aria-label="anchor" href="#functions-and-utilities"></a>
</h2>
<table class="table">
<colgroup>
<col width="9%">
<col width="90%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Functions available</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>SpatialExperiment</code></td>
<td>All</td>
</tr>
<tr class="even">
<td><code>dplyr</code></td>
<td>
<code>arrange</code>,<code>bind_rows</code>, <code>bind_cols</code>, <code>distinct</code>, <code>filter</code>, <code>group_by</code>, <code>summarise</code>, <code>select</code>, <code>mutate</code>, <code>rename</code>, <code>left_join</code>, <code>right_join</code>, <code>inner_join</code>, <code>slice</code>, <code>sample_n</code>, <code>sample_frac</code>, <code>count</code>, <code>add_count</code>
</td>
</tr>
<tr class="odd">
<td><code>tidyr</code></td>
<td>
<code>nest</code>, <code>unnest</code>, <code>unite</code>, <code>separate</code>, <code>extract</code>, <code>pivot_longer</code>
</td>
</tr>
<tr class="even">
<td><code>ggplot2</code></td>
<td><code>ggplot</code></td>
</tr>
<tr class="odd">
<td><code>plotly</code></td>
<td><code>plot_ly</code></td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="20%">
<col width="79%">
</colgroup>
<thead><tr class="header">
<th>Utility</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
<tr class="even">
<td><code>join_features</code></td>
<td>Combine cell-wise and feature-wise information into a <code>tbl_df</code>
</td>
</tr>
<tr class="odd">
<td><code>aggregate_cells</code></td>
<td>Aggregate cell-feature abundance into a pseudobulk <code>SummarizedExperiment</code>
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="spatialexperiment-tibble-abstraction">SpatialExperiment-tibble abstraction<a class="anchor" aria-label="anchor" href="#spatialexperiment-tibble-abstraction"></a>
</h2>
<p>A SpatialExperiment object represents observations (cells) as columns and variables (features) as rows, as is the Bioconductor convention. Additional information about the cells is accessed through the <code>reducedDims</code>, <code>colData</code> and <code>spatialCoords</code> functions.</p>
<p>tidySpatialExperiment provides a SpatialExperiment-tibble abstraction, representing cells as rows and features as columns, as is the <em>tidyverse</em> convention. <code>colData</code> and <code>spatialCoords</code> are appended as columns to the same abstraction, allowing easy interaction with this additional data.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of tidySpatialExperiment from GitHub with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"william-hutchison/tidySpatialExperiment"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>Here, we load an example SpatialExperiment object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example SpatialExperiment object</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/william-hutchison/tidySpatialExperiment" class="external-link">tidySpatialExperiment</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">read10xVisium</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="view-data">View data<a class="anchor" aria-label="anchor" href="#view-data"></a>
</h2>
<p>The default view is now of the SpatialExperiment-tibble abstraction.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 99 Ã— 7</span></span>
<span><span class="co">#  # [90mFeatures=50 | Cells=99 | Assays=counts[0m</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                7679</span></span>
<span><span class="co">#  # â„¹ 94 more rows</span></span>
<span><span class="co">#  # â„¹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
<p>However, our data maintains its status as a SpatialExperiment object. Therefore, we have access to all SpatialExperiment functions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">#                     in_tissue array_row array_col   sample_id</span></span>
<span><span class="co">#                     &lt;logical&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">#  AAACAACGAATAGTTC-1     FALSE         0        16    section1</span></span>
<span><span class="co">#  AAACAAGTATCTCCCA-1      TRUE        50       102    section1</span></span>
<span><span class="co">#  AAACAATCTACTAGCA-1      TRUE         3        43    section1</span></span>
<span><span class="co">#  AAACACCAATAACTGC-1      TRUE        59        19    section1</span></span>
<span><span class="co">#  AAACAGAGCGACTCCT-1      TRUE        14        94    section1</span></span>
<span><span class="co">#  AAACAGCTTTCAGAAG-1     FALSE        43         9    section1</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#                     pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">#  AAACAACGAATAGTTC-1               2312               1252</span></span>
<span><span class="co">#  AAACAAGTATCTCCCA-1               8230               7237</span></span>
<span><span class="co">#  AAACAATCTACTAGCA-1               4170               1611</span></span>
<span><span class="co">#  AAACACCAATAACTGC-1               2519               8315</span></span>
<span><span class="co">#  AAACAGAGCGACTCCT-1               7679               2927</span></span>
<span><span class="co">#  AAACAGCTTTCAGAAG-1               1831               6400</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  DataFrame with 2 rows and 4 columns</span></span>
<span><span class="co">#      sample_id    image_id   data scaleFactor</span></span>
<span><span class="co">#    &lt;character&gt; &lt;character&gt; &lt;list&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#  1    section1      lowres   ####   0.0510334</span></span>
<span><span class="co">#  2    section2      lowres   ####   0.0510334</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="integration-with-the-tidyverse-ecosystem">Integration with the <em>tidyverse</em> ecosystem<a class="anchor" aria-label="anchor" href="#integration-with-the-tidyverse-ecosystem"></a>
</h1>
<div class="section level2">
<h2 id="manipulate-with-dplyr">Manipulate with dplyr<a class="anchor" aria-label="anchor" href="#manipulate-with-dplyr"></a>
</h2>
<p>Most functions from dplyr are available for use with the SpatialExperiment-tibble abstraction. For example, <code>filter</code> can be used to select cells by a variable of interest.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">array_col</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 6 Ã— 7</span></span>
<span><span class="co">#  # [90mFeatures=50 | Cells=6 | Assays=counts[0m</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACATGGTGAGAGGA-1 FALSE            62         0 section1                1212</span></span>
<span><span class="co">#  2 AAACGAAGATGGAGTA-1 FALSE            58         4 section1                1487</span></span>
<span><span class="co">#  3 AAAGAATGACCTTAGA-1 FALSE            64         2 section1                1349</span></span>
<span><span class="co">#  4 AAACATGGTGAGAGGA-1 FALSE            62         0 section2                1212</span></span>
<span><span class="co">#  5 AAACGAAGATGGAGTA-1 FALSE            58         4 section2                1487</span></span>
<span><span class="co">#  # â„¹ 1 more row</span></span>
<span><span class="co">#  # â„¹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
<p>And <code>mutate</code> can be used to add new variables, or modify the value of an existing variable.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>in_region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">&amp;</span> <span class="va">array_row</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 99 Ã— 8</span></span>
<span><span class="co">#  # [90mFeatures=50 | Cells=99 | Assays=counts[0m</span></span>
<span><span class="co">#    .cell     in_tissue array_row array_col sample_id in_region pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;     &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;     &lt;lgl&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGâ€¦ FALSE             0        16 section1  FALSE                   2312</span></span>
<span><span class="co">#  2 AAACAAGTâ€¦ TRUE             50       102 section1  FALSE                   8230</span></span>
<span><span class="co">#  3 AAACAATCâ€¦ TRUE              3        43 section1  TRUE                    4170</span></span>
<span><span class="co">#  4 AAACACCAâ€¦ TRUE             59        19 section1  FALSE                   2519</span></span>
<span><span class="co">#  5 AAACAGAGâ€¦ TRUE             14        94 section1  FALSE                   7679</span></span>
<span><span class="co">#  # â„¹ 94 more rows</span></span>
<span><span class="co">#  # â„¹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tidy-with-tidyr">Tidy with tidyr<a class="anchor" aria-label="anchor" href="#tidy-with-tidyr"></a>
</h2>
<p>Most functions from tidyr are also available. Here, <code>nest</code> is used to group the data by <code>sample_id</code>, and <code>unnest</code> is used to ungroup the data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Nest the SpatialExperiment object by sample_id</span></span>
<span><span class="va">spe_nested</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/nest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the nested SpatialExperiment object</span></span>
<span><span class="va">spe_nested</span></span>
<span><span class="co">#  # A tibble: 2 Ã— 2</span></span>
<span><span class="co">#    sample_id data           </span></span>
<span><span class="co">#    &lt;chr&gt;     &lt;list&gt;         </span></span>
<span><span class="co">#  1 section1  &lt;SptlExpr[,50]&gt;</span></span>
<span><span class="co">#  2 section2  &lt;SptlExpr[,49]&gt;</span></span>
<span></span>
<span><span class="co"># Unnest the nested SpatialExperiment objects</span></span>
<span><span class="va">spe_nested</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 99 Ã— 7</span></span>
<span><span class="co">#  # [90mFeatures=50 | Cells=99 | Assays=counts[0m</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                7679</span></span>
<span><span class="co">#  # â„¹ 94 more rows</span></span>
<span><span class="co">#  # â„¹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="plot-with-ggplot2">Plot with ggplot2<a class="anchor" aria-label="anchor" href="#plot-with-ggplot2"></a>
</h2>
<p>The <code>ggplot</code> function can be used to create a plot from a SpatialExperiment object. This example also demonstrates how tidy operations can be combined to build up more complex analysis. It should be noted that helper functions such <code>aes</code> are not included and should be imported from ggplot2.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span> <span class="op">&amp;</span> <span class="va">in_tissue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Add a column with the sum of feature counts per cell</span></span>
<span>  <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>count_sum <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_int</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="op">~</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>, <span class="va">.x</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Plot with tidySpatialExperiment and ggplot2</span></span>
<span>  <span class="fu"><a href="reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">count_sum</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">count_sum</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/unnamed-chunk-10-1.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="plot-with-plotly">Plot with plotly<a class="anchor" aria-label="anchor" href="#plot-with-plotly"></a>
</h2>
<p>The <code>plot_ly</code> function can also be used to create a plot from a SpatialExperiment object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/plot_ly.html">plot_ly</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="op">~</span> <span class="va">array_col</span>, </span>
<span>    y <span class="op">=</span> <span class="op">~</span> <span class="va">array_row</span>, </span>
<span>    color <span class="op">=</span> <span class="op">~</span> <span class="va">in_tissue</span>, </span>
<span>    type <span class="op">=</span> <span class="st">"scatter"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/plotly_demo.png"></p>
</div>
</div>
<div class="section level1">
<h1 id="integration-with-the-tidyomics-ecosystem">Integration with the <em>tidyomics</em> ecosystem<a class="anchor" aria-label="anchor" href="#integration-with-the-tidyomics-ecosystem"></a>
</h1>
<div class="section level2">
<h2 id="interactively-select-cells-with-tidygate">Interactively select cells with tidygate<a class="anchor" aria-label="anchor" href="#interactively-select-cells-with-tidygate"></a>
</h2>
<p>Different packages from the <em>tidyomics</em> ecosystem are easy to use together. Here, tidygate is used to interactively gate cells based on their array location.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_regions</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>region <span class="op">=</span> <span class="fu">tidygate</span><span class="fu">::</span><span class="fu">gate_chr</span><span class="op">(</span><span class="va">array_col</span>, <span class="va">array_row</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/tidygate_demo.gif"></p>
<p>The gated cells can then be divided into pseudobulks within a SummarizedExperiment object using tidySpatialExperimentâ€™s <code>aggregate_cells</code> utility function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_regions_aggregated</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">spe_regions</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/aggregate_cells.html">aggregate_cells</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="important-considerations">Important considerations<a class="anchor" aria-label="anchor" href="#important-considerations"></a>
</h1>
<div class="section level2">
<h2 id="read-only-columns">Read-only columns<a class="anchor" aria-label="anchor" href="#read-only-columns"></a>
</h2>
<p>Removing the <code>.cell</code> column will return a tibble. This is consistent with the behaviour in other <em>tidyomics</em> packages.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">.cell</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  tidySpatialExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span></span>
<span><span class="co">#  # A tibble: 6 Ã— 4</span></span>
<span><span class="co">#    in_tissue array_row array_col sample_id</span></span>
<span><span class="co">#    &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;    </span></span>
<span><span class="co">#  1 FALSE             0        16 section1 </span></span>
<span><span class="co">#  2 TRUE             50       102 section1 </span></span>
<span><span class="co">#  3 TRUE              3        43 section1 </span></span>
<span><span class="co">#  4 TRUE             59        19 section1 </span></span>
<span><span class="co">#  5 TRUE             14        94 section1 </span></span>
<span><span class="co">#  # â„¹ 1 more row</span></span></code></pre></div>
<p>The sample_id column cannot be removed with <em>tidyverse</em> functions, and can only be modified if the changes are accepted by SpatialExperimentâ€™s <code>colData</code> function.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample_id is not removed, despite the user's request</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample_id</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 99 Ã— 7</span></span>
<span><span class="co">#  # [90mFeatures=50 | Cells=99 | Assays=counts[0m</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                  &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1                2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1                8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1                4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1                2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1                7679</span></span>
<span><span class="co">#  # â„¹ 94 more rows</span></span>
<span><span class="co">#  # â„¹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span>
<span></span>
<span><span class="co"># This change maintains separation of sample_ids and is permitted</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="st">"_modified"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#  # A SpatialExperiment-tibble abstraction: 99 Ã— 7</span></span>
<span><span class="co">#  # [90mFeatures=6 | Cells=99 | Assays=counts[0m</span></span>
<span><span class="co">#    .cell              in_tissue array_row array_col sample_id  pxl_col_in_fullres</span></span>
<span><span class="co">#    &lt;chr&gt;              &lt;lgl&gt;         &lt;int&gt;     &lt;int&gt; &lt;chr&gt;                   &lt;int&gt;</span></span>
<span><span class="co">#  1 AAACAACGAATAGTTC-1 FALSE             0        16 section1_â€¦               2312</span></span>
<span><span class="co">#  2 AAACAAGTATCTCCCA-1 TRUE             50       102 section1_â€¦               8230</span></span>
<span><span class="co">#  3 AAACAATCTACTAGCA-1 TRUE              3        43 section1_â€¦               4170</span></span>
<span><span class="co">#  4 AAACACCAATAACTGC-1 TRUE             59        19 section1_â€¦               2519</span></span>
<span><span class="co">#  5 AAACAGAGCGACTCCT-1 TRUE             14        94 section1_â€¦               7679</span></span>
<span><span class="co">#  # â„¹ 94 more rows</span></span>
<span><span class="co">#  # â„¹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span>
<span></span>
<span><span class="co"># This change does not maintain separation of sample_ids and produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="st">"new_sample"</span><span class="op">)</span></span>
<span><span class="co">#  Error in .local(x, ..., value): Number of unique 'sample_id's is 2, but 1 was provided.</span></span></code></pre></div>
<p>The <code>pxl_col_in_fullres</code> and <code>px_row_in_fullres</code> columns cannot be removed or modified with <em>tidyverse</em> functions. This is consistent with the behaviour of dimension reduction data in other <em>tidyomics</em> packages.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attempting to remove pxl_col_in_fullres produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">pxl_col_in_fullres</span><span class="op">)</span></span>
<span><span class="co">#  Error in `select_helper()`:</span></span>
<span><span class="co">#  ! Can't subset columns that don't exist.</span></span>
<span><span class="co">#  âœ– Column `pxl_col_in_fullres` doesn't exist.</span></span>
<span></span>
<span><span class="co"># Attempting to modify pxl_col_in_fullres produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">pxl_col_in_fullres</span><span class="op">)</span></span>
<span><span class="co">#  Error in `dplyr::mutate()`:</span></span>
<span><span class="co">#  â„¹ In argument: `pxl_col_in_fullres`.</span></span>
<span><span class="co">#  Caused by error:</span></span>
<span><span class="co">#  ! object 'pxl_col_in_fullres' not found</span></span></code></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/william-hutchison/tidySpatialExperiment/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/william-hutchison/tidySpatialExperiment/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidySpatialExperiment</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>William Hutchison <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0009-0001-6242-4269" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Stefano Mangiola <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-blue.svg" alt="Lifecycle:experimental"></a></li>
<li><a href="https://github.com/william-hutchison/tidySpatialExperiment/actions" class="external-link"><img src="https://github.com/stemangiola/tidybulk/workflows/rworkflows/badge.svg" alt="R build status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by William Hutchison, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
