<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview • tidySpatialExperiment</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview">
<meta property="og:description" content="tidySpatialExperiment">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidySpatialExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/overview.html">Overview</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/william-hutchison/tidySpatialExperiment/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview</h1>
                        <h4 data-toc-skip class="author">William
Hutchison</h4>
            <address class="author_afil">
      WEHI - Walter and Eliza Hall Institute of Medical
Research<br><a class="author_email" href="mailto:#"></a><a href="mailto:hutchison.w@wehi.edu.au" class="email">hutchison.w@wehi.edu.au</a>
      </address>
                              <h4 data-toc-skip class="author">Stefano
Mangiola</h4>
            <address class="author_afil">
      WEHI - Walter and Eliza Hall Institute of Medical
Research<br><small class="dont-index">Source: <a href="https://github.com/william-hutchison/tidySpatialExperiment/blob/HEAD/vignettes/overview.Rmd" class="external-link"><code>vignettes/overview.Rmd</code></a></small>
      <div class="hidden name"><code>overview.Rmd</code></div>

    </address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>A brief overview of the tidySpatialExperiment package -
      demonstrating the SpatialExperiment-tibble abstraction,
      compatibility with the <em>tidyverse</em> ecosystem, compatibility
      with the <em>tidyomics</em> ecosystem and a few helpful utility
      functions.</p>
    </div>
    
<div class="section level2">
<h2 id="tidyspatialexperiment---part-of-tidyomics">tidySpatialExperiment - part of <em>tidyomics</em><a class="anchor" aria-label="anchor" href="#tidyspatialexperiment---part-of-tidyomics"></a>
</h2>
<p>Resources to help you get started with tidySpatialExperiment and
<em>tidyomics</em>:</p>
<ul>
<li><a href="http://william-hutchison.github.io/tidySpatialExperiment/" class="external-link">The
tidySpatialExperiment website</a></li>
<li><a href="https://tidyomics.github.io/tidyomicsBlog/" class="external-link">The tidyomics
blog</a></li>
<li><a href="https://rstudio-pubs-static.s3.amazonaws.com/792462_f948e766b15d4ee5be5c860493bda0b3.html" class="external-link">Third
party tutorials</a></li>
</ul>
<p>The <em>tidyomics</em> ecosystem includes packages for:</p>
<ul>
<li>
<p>Working with genomic features:</p>
<ul>
<li>
<a href="https://github.com/sa-lee/plyranges" class="external-link">plyranges</a>, for
tidy manipulation of genomic range data.</li>
<li>
<a href="https://github.com/nullranges/nullranges" class="external-link">nullranges</a>,
for tidy generation of genomic ranges representing the null
hypothesis.</li>
<li>
<a href="https://github.com/tidyomics/plyinteractions" class="external-link">plyinteractions</a>,
for tidy manipulation of genomic interaction data.</li>
</ul>
</li>
<li>
<p>Working with transcriptomic features:</p>
<ul>
<li>
<a href="https://github.com/stemangiola/tidySummarizedExperiment" class="external-link">tidySummarizedExperiment</a>,
for tidy manipulation of SummarizedExperiment objects.</li>
<li>
<a href="https://github.com/stemangiola/tidySingleCellExperiment" class="external-link">tidySingleCellExperiment</a>,
for tidy manipulation of SingleCellExperiment objects.</li>
<li>
<a href="https://github.com/stemangiola/tidyseurat" class="external-link">tidyseurat</a>,
for tidy manipulation of Seurat objects.</li>
<li>
<a href="https://github.com/stemangiola/tidybulk" class="external-link">tidybulk</a>, for
bulk RNA-seq analysis.</li>
</ul>
</li>
<li>
<p>Working with cytometry features:</p>
<ul>
<li>
<a href="https://github.com/keyes-timothy/tidytof" class="external-link">tidytof</a>, for
tidy manipulation of high-dimensional cytometry data.</li>
</ul>
</li>
</ul>
<p>A few more tidy tools for data manipulation and plotting:</p>
<ul>
<li>
<a href="https://github.com/stemangiola/tidyHeatmap" class="external-link">tidyHeatmap</a>, for
producing heatmaps with tidy principles. analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate" class="external-link">tidygate</a>, for
interactive plotting and gating.</li>
</ul>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>tidySpatialExperiment provides a bridge between the <a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a>
<span class="citation">[@righelli2022spatialexperiment]</span> package
and the <a href="https://www.tidyverse.org" class="external-link"><em>tidyverse</em></a> <span class="citation">[@wickham2019welcome]</span> ecosystem. It creates an
invisible layer that allows you to interact with a SpatialExperiment
object as if it were a tibble; enabling the use of functions from <a href="https://github.com/tidyverse/dplyr" class="external-link">dplyr</a>, <a href="https://github.com/tidyverse/tidyr" class="external-link">tidyr</a>, <a href="https://github.com/tidyverse/ggplot2" class="external-link">ggplot2</a> and <a href="https://github.com/plotly/plotly.R" class="external-link">plotly</a>. But, underneath,
your data remains a SpatialExperiment object.</p>
<p>tidySpatialExperiment also provides three additional utility
functions.</p>
<div class="section level3">
<h3 id="functions-and-utilities">Functions and utilities<a class="anchor" aria-label="anchor" href="#functions-and-utilities"></a>
</h3>
<table class="table">
<colgroup>
<col width="48%">
<col width="51%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Functions available</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>SpatialExperiment</code></td>
<td>All</td>
</tr>
<tr class="even">
<td><code>dplyr</code></td>
<td>
<code>arrange</code>,<code>bind_rows</code>, <code>bind_cols</code>,
<code>distinct</code>, <code>filter</code>, <code>group_by</code>,
<code>summarise</code>, <code>select</code>, <code>mutate</code>,
<code>rename</code>, <code>left_join</code>, <code>right_join</code>,
<code>inner_join</code>, <code>slice</code>, <code>sample_n</code>,
<code>sample_frac</code>, <code>count</code>,
<code>add_count</code>
</td>
</tr>
<tr class="odd">
<td><code>tidyr</code></td>
<td>
<code>nest</code>, <code>unnest</code>, <code>unite</code>,
<code>separate</code>, <code>extract</code>,
<code>pivot_longer</code>
</td>
</tr>
<tr class="even">
<td><code>ggplot2</code></td>
<td><code>ggplot</code></td>
</tr>
<tr class="odd">
<td><code>plotly</code></td>
<td><code>plot_ly</code></td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="48%">
<col width="51%">
</colgroup>
<thead><tr class="header">
<th>Utility</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>as_tibble</code></td>
<td>Convert cell data to a <code>tbl_df</code>
</td>
</tr>
<tr class="even">
<td><code>join_features</code></td>
<td>Append feature data to cell data</td>
</tr>
<tr class="odd">
<td><code>aggregate_cells</code></td>
<td>Aggregate cell-feature abundance into a pseudobulk
<code>SummarizedExperiment</code> object</td>
</tr>
<tr class="even">
<td><code>rectangle</code></td>
<td>Select rectangular region of space</td>
</tr>
<tr class="odd">
<td><code>ellipse</code></td>
<td>Select elliptical region of space</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>You can install the development version of tidySpatialExperiment from
GitHub with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"william-hutchison/tidySpatialExperiment"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>Here, we load an example SpatialExperiment object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load example SpatialExperiment object</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/william-hutchison/tidySpatialExperiment" class="external-link">tidySpatialExperiment</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'tidySingleCellExperiment::plot_ly' by</span></span>
<span><span class="co">## 'ttservice::plot_ly' when loading 'tidySpatialExperiment'</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/example.html" class="external-link">example</a></span><span class="op">(</span><span class="va">read10xVisium</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="spatialexperiment-tibble-abstraction">SpatialExperiment-tibble abstraction<a class="anchor" aria-label="anchor" href="#spatialexperiment-tibble-abstraction"></a>
</h3>
<p>A SpatialExperiment object represents observations (cells) as columns
and variables (features) as rows, as is the Bioconductor convention.
Additional information about the cells is accessed through the
<code>reducedDims</code>, <code>colData</code> and
<code>spatialCoords</code> functions.</p>
<p>tidySpatialExperiment provides a SpatialExperiment-tibble
abstraction, representing cells as rows and features as columns, as is
the <em>tidyverse</em> convention. <code>colData</code> and
<code>spatialCoords</code> are appended as columns to the same
abstraction, allowing easy interaction with this additional data.</p>
<p>The default view is now of the SpatialExperiment-tibble
abstraction.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 99 × 7</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=50 | Cells=99 | Assays=counts</span></span></span>
<span><span class="co">##    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAACGAATAGTTC-1 FALSE             0        16 section1                <span style="text-decoration: underline;">2</span>312</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAAGTATCTCCCA-1 TRUE             50       102 section1                <span style="text-decoration: underline;">8</span>230</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAATCTACTAGCA-1 TRUE              3        43 section1                <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACACCAATAACTGC-1 TRUE             59        19 section1                <span style="text-decoration: underline;">2</span>519</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGAGCGACTCCT-1 TRUE             14        94 section1                <span style="text-decoration: underline;">7</span>679</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGCTTTCAGAAG-1 FALSE            43         9 section1                <span style="text-decoration: underline;">1</span>831</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACAGGGTCTATATT-1 FALSE            47        13 section1                <span style="text-decoration: underline;">2</span>106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACAGTGTTCCTGGG-1 FALSE            73        43 section1                <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section1                <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACATTTCCCGGATT-1 FALSE            61        97 section1                <span style="text-decoration: underline;">7</span>886</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 89 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
<p>However, our data maintains its status as a SpatialExperiment object.
Therefore, we have access to all SpatialExperiment functions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##                    in_tissue array_row array_col   sample_id</span></span>
<span><span class="co">##                    &lt;logical&gt; &lt;integer&gt; &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">## AAACAACGAATAGTTC-1     FALSE         0        16    section1</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1      TRUE        50       102    section1</span></span>
<span><span class="co">## AAACAATCTACTAGCA-1      TRUE         3        43    section1</span></span>
<span><span class="co">## AAACACCAATAACTGC-1      TRUE        59        19    section1</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1      TRUE        14        94    section1</span></span>
<span><span class="co">## AAACAGCTTTCAGAAG-1     FALSE        43         9    section1</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    pxl_col_in_fullres pxl_row_in_fullres</span></span>
<span><span class="co">## AAACAACGAATAGTTC-1               2312               1252</span></span>
<span><span class="co">## AAACAAGTATCTCCCA-1               8230               7237</span></span>
<span><span class="co">## AAACAATCTACTAGCA-1               4170               1611</span></span>
<span><span class="co">## AAACACCAATAACTGC-1               2519               8315</span></span>
<span><span class="co">## AAACAGAGCGACTCCT-1               7679               2927</span></span>
<span><span class="co">## AAACAGCTTTCAGAAG-1               1831               6400</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">imgData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 2 rows and 4 columns</span></span>
<span><span class="co">##     sample_id    image_id   data scaleFactor</span></span>
<span><span class="co">##   &lt;character&gt; &lt;character&gt; &lt;list&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1    section1      lowres   ####   0.0510334</span></span>
<span><span class="co">## 2    section2      lowres   ####   0.0510334</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="integration-with-the-tidyverse-ecosystem">Integration with the <em>tidyverse</em> ecosystem<a class="anchor" aria-label="anchor" href="#integration-with-the-tidyverse-ecosystem"></a>
</h2>
<div class="section level3">
<h3 id="manipulate-with-dplyr">Manipulate with dplyr<a class="anchor" aria-label="anchor" href="#manipulate-with-dplyr"></a>
</h3>
<p>Most functions from dplyr are available for use with the
SpatialExperiment-tibble abstraction. For example, <code>filter</code>
can be used to select cells by a variable of interest.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">array_col</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 6 × 7</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=50 | Cells=6 | Assays=counts</span></span></span>
<span><span class="co">##   .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section1                <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> AAACGAAGATGGAGTA-1 FALSE            58         4 section1                <span style="text-decoration: underline;">1</span>487</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> AAAGAATGACCTTAGA-1 FALSE            64         2 section1                <span style="text-decoration: underline;">1</span>349</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section2                <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> AAACGAAGATGGAGTA-1 FALSE            58         4 section2                <span style="text-decoration: underline;">1</span>487</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> AAAGAATGACCTTAGA-1 FALSE            64         2 section2                <span style="text-decoration: underline;">1</span>349</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
<p>And <code>mutate</code> can be used to add new variables, or modify
the value of an existing variable.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span>in_region <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">in_tissue</span> <span class="op">&amp;</span> <span class="va">array_row</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 99 × 8</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=50 | Cells=99 | Assays=counts</span></span></span>
<span><span class="co">##    .cell    in_tissue array_row array_col sample_id in_region pxl_col_in_fullres</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAAC… FALSE             0        16 section1  FALSE                   <span style="text-decoration: underline;">2</span>312</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAAG… TRUE             50       102 section1  FALSE                   <span style="text-decoration: underline;">8</span>230</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAAT… TRUE              3        43 section1  TRUE                    <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACACC… TRUE             59        19 section1  FALSE                   <span style="text-decoration: underline;">2</span>519</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGA… TRUE             14        94 section1  FALSE                   <span style="text-decoration: underline;">7</span>679</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGC… FALSE            43         9 section1  FALSE                   <span style="text-decoration: underline;">1</span>831</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACAGG… FALSE            47        13 section1  FALSE                   <span style="text-decoration: underline;">2</span>106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACAGT… FALSE            73        43 section1  FALSE                   <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACATG… FALSE            62         0 section1  FALSE                   <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACATT… FALSE            61        97 section1  FALSE                   <span style="text-decoration: underline;">7</span>886</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 89 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="tidy-with-tidyr">Tidy with tidyr<a class="anchor" aria-label="anchor" href="#tidy-with-tidyr"></a>
</h3>
<p>Most functions from tidyr are also available. Here, <code>nest</code>
is used to group the data by <code>sample_id</code>, and
<code>unnest</code> is used to ungroup the data.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Nest the SpatialExperiment object by sample_id</span></span>
<span><span class="va">spe_nested</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/nest.html">nest</a></span><span class="op">(</span>data <span class="op">=</span> <span class="op">-</span><span class="va">sample_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View the nested SpatialExperiment object</span></span>
<span><span class="va">spe_nested</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   sample_id data           </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> section1  <span style="color: #949494;">&lt;SptlExpr[,50]&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> section2  <span style="color: #949494;">&lt;SptlExpr[,49]&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Unnest the nested SpatialExperiment objects</span></span>
<span><span class="va">spe_nested</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/unnest.html">unnest</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 99 × 7</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=50 | Cells=99 | Assays=counts</span></span></span>
<span><span class="co">##    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAACGAATAGTTC-1 FALSE             0        16 section1                <span style="text-decoration: underline;">2</span>312</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAAGTATCTCCCA-1 TRUE             50       102 section1                <span style="text-decoration: underline;">8</span>230</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAATCTACTAGCA-1 TRUE              3        43 section1                <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACACCAATAACTGC-1 TRUE             59        19 section1                <span style="text-decoration: underline;">2</span>519</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGAGCGACTCCT-1 TRUE             14        94 section1                <span style="text-decoration: underline;">7</span>679</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGCTTTCAGAAG-1 FALSE            43         9 section1                <span style="text-decoration: underline;">1</span>831</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACAGGGTCTATATT-1 FALSE            47        13 section1                <span style="text-decoration: underline;">2</span>106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACAGTGTTCCTGGG-1 FALSE            73        43 section1                <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section1                <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACATTTCCCGGATT-1 FALSE            61        97 section1                <span style="text-decoration: underline;">7</span>886</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 89 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="plot-with-ggplot2">Plot with ggplot2<a class="anchor" aria-label="anchor" href="#plot-with-ggplot2"></a>
</h3>
<p>The <code>ggplot</code> function can be used to create a plot from a
SpatialExperiment object. This example also demonstrates how tidy
operations can be combined to build up more complex analysis. It should
be noted that helper functions such <code>aes</code> are not included
and should be imported from ggplot2.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span> <span class="op">&amp;</span> <span class="va">in_tissue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Add a column with the sum of feature counts per cell</span></span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span>count_sum <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map_int</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="op">~</span></span>
<span>    <span class="va">spe</span><span class="op">[</span>, <span class="va">.x</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu">counts</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  </span>
<span>  <span class="co"># Plot with tidySpatialExperiment and ggplot2</span></span>
<span>  <span class="fu"><a href="../reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">count_sum</span><span class="op">)</span>, y <span class="op">=</span> <span class="va">count_sum</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="plot-with-plotly">Plot with plotly<a class="anchor" aria-label="anchor" href="#plot-with-plotly"></a>
</h3>
<p>The <code>plot_ly</code> function can also be used to create a plot
from a SpatialExperiment object.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/plot_ly.html">plot_ly</a></span><span class="op">(</span></span>
<span>    x <span class="op">=</span> <span class="op">~</span> <span class="va">array_col</span>, </span>
<span>    y <span class="op">=</span> <span class="op">~</span> <span class="va">array_row</span>, </span>
<span>    color <span class="op">=</span> <span class="op">~</span> <span class="va">in_tissue</span>, </span>
<span>    type <span class="op">=</span> <span class="st">"scatter"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../reference/figures/plotly_demo.png" alt=""><p class="caption">plotly demonstration</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="integration-with-the-tidyomics-ecosystem">Integration with the <em>tidyomics</em> ecosystem<a class="anchor" aria-label="anchor" href="#integration-with-the-tidyomics-ecosystem"></a>
</h2>
<div class="section level3">
<h3 id="interactively-select-cells-with-tidygate">Interactively select cells with tidygate<a class="anchor" aria-label="anchor" href="#interactively-select-cells-with-tidygate"></a>
</h3>
<p>Different packages from the <em>tidyomics</em> ecosystem are easy to
use together. Here, tidygate is used to interactively gate cells based
on their array location.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_regions</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span>region <span class="op">=</span> <span class="fu">tidygate</span><span class="fu">::</span><span class="fu">gate_chr</span><span class="op">(</span><span class="va">array_col</span>, <span class="va">array_row</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="../reference/figures/tidygate_demo.gif" alt=""><p class="caption">tidygate demonstration</p>
</div>
<p>The gated cells can then be divided into pseudobulks within a
SummarizedExperiment object using tidySpatialExperiment’s
<code>aggregate_cells</code> utility function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe_regions_aggregated</span> <span class="op">&lt;-</span></span>
<span>  <span class="va">spe_regions</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/aggregate_cells.html">aggregate_cells</a></span><span class="op">(</span><span class="va">region</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a>
</h2>
<div class="section level3">
<h3 id="append-feature-data-to-cell-data">Append feature data to cell data<a class="anchor" aria-label="anchor" href="#append-feature-data-to-cell-data"></a>
</h3>
<p>The <em>tidyomics</em> ecosystem places the emphasis on interacting
with cell data. To interact with feature data, the
<code>join_feature</code> function can be used to append feature values
to cell data.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Join feature data in wide format, preserving the SpatialExperiment object</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSMUSG00000025915"</span>, <span class="st">"ENSMUSG00000042501"</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 99 × 9</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=6 | Cells=99 | Assays=counts</span></span></span>
<span><span class="co">##    .cell              in_tissue array_row array_col sample_id ENSMUSG00000025915</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAACGAATAGTTC-1 FALSE             0        16 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAAGTATCTCCCA-1 TRUE             50       102 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAATCTACTAGCA-1 TRUE              3        43 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACACCAATAACTGC-1 TRUE             59        19 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGAGCGACTCCT-1 TRUE             14        94 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGCTTTCAGAAG-1 FALSE            43         9 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACAGGGTCTATATT-1 FALSE            47        13 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACAGTGTTCCTGGG-1 FALSE            73        43 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section1                   0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACATTTCCCGGATT-1 FALSE            61        97 section1                   0</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 89 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: ENSMUSG00000042501 &lt;dbl&gt;, pxl_col_in_fullres &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Join feature data in long format, discarding the SpatialExperiment object</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/join_features.html">join_features</a></span><span class="op">(</span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSMUSG00000025915"</span>, <span class="st">"ENSMUSG00000042501"</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"long"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySpatialExperiment says: A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 7</span></span></span>
<span><span class="co">##   .cell       in_tissue array_row array_col sample_id .feature .abundance_counts</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> AAACAACGAA… FALSE             0        16 section1  ENSMUSG…                 0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> AAACAACGAA… FALSE             0        16 section1  ENSMUSG…                 0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> AAACAAGTAT… TRUE             50       102 section1  ENSMUSG…                 0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> AAACAAGTAT… TRUE             50       102 section1  ENSMUSG…                 1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> AAACAATCTA… TRUE              3        43 section1  ENSMUSG…                 0</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> AAACAATCTA… TRUE              3        43 section1  ENSMUSG…                 0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="aggregate-cells">Aggregate cells<a class="anchor" aria-label="anchor" href="#aggregate-cells"></a>
</h3>
<p>Sometimes, it is necessary to aggregate the gene-transcript abundance
from a group of cells into a single value. For example, when comparing
groups of cells across different samples with fixed-effect models.</p>
<p>Cell aggregation can be achieved using the
<code>aggregate_cells</code> function.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/aggregate_cells.html">aggregate_cells</a></span><span class="op">(</span><span class="va">in_tissue</span>, assays <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SummarizedExperiment </span></span>
<span><span class="co">## dim: 50 2 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(50): ENSMUSG00000002459 ENSMUSG00000005886 ...</span></span>
<span><span class="co">##   ENSMUSG00000104217 ENSMUSG00000104328</span></span>
<span><span class="co">## rowData names(1): feature</span></span>
<span><span class="co">## colnames(2): FALSE TRUE</span></span>
<span><span class="co">## colData names(2): in_tissue .aggregated_cells</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="elliptical-and-rectangular-region-selection">Elliptical and rectangular region selection<a class="anchor" aria-label="anchor" href="#elliptical-and-rectangular-region-selection"></a>
</h3>
<p>To select cells by their geometric region in space, the
<code>ellipse</code> and <code>rectangle</code> functions can be
used.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/filter.html">filter</a></span><span class="op">(</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"section1"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span>in_ellipse <span class="op">=</span> <span class="fu"><a href="../reference/ellipse.html">ellipse</a></span><span class="op">(</span><span class="va">array_col</span>, <span class="va">array_row</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">40</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">array_col</span>, y <span class="op">=</span> <span class="va">array_row</span>, colour <span class="op">=</span> <span class="va">in_ellipse</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="important-considerations">Important considerations<a class="anchor" aria-label="anchor" href="#important-considerations"></a>
</h2>
<div class="section level3">
<h3 id="read-only-columns">Read-only columns<a class="anchor" aria-label="anchor" href="#read-only-columns"></a>
</h3>
<p>Removing the <code>.cell</code> column will return a tibble. This is
consistent with the behaviour in other <em>tidyomics</em> packages.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">.cell</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## tidySpatialExperiment says: Key columns are missing. A data frame is returned for independent data analysis.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   in_tissue array_row array_col sample_id</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> FALSE             0        16 section1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> TRUE             50       102 section1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> TRUE              3        43 section1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> TRUE             59        19 section1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> TRUE             14        94 section1 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> FALSE            43         9 section1</span></span></code></pre>
<p>The sample_id column cannot be removed with <em>tidyverse</em>
functions, and can only be modified if the changes are accepted by
SpatialExperiment’s <code>colData</code> function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample_id is not removed, despite the user's request</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">sample_id</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 99 × 7</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=50 | Cells=99 | Assays=counts</span></span></span>
<span><span class="co">##    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAACGAATAGTTC-1 FALSE             0        16 section1                <span style="text-decoration: underline;">2</span>312</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAAGTATCTCCCA-1 TRUE             50       102 section1                <span style="text-decoration: underline;">8</span>230</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAATCTACTAGCA-1 TRUE              3        43 section1                <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACACCAATAACTGC-1 TRUE             59        19 section1                <span style="text-decoration: underline;">2</span>519</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGAGCGACTCCT-1 TRUE             14        94 section1                <span style="text-decoration: underline;">7</span>679</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGCTTTCAGAAG-1 FALSE            43         9 section1                <span style="text-decoration: underline;">1</span>831</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACAGGGTCTATATT-1 FALSE            47        13 section1                <span style="text-decoration: underline;">2</span>106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACAGTGTTCCTGGG-1 FALSE            73        43 section1                <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section1                <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACATTTCCCGGATT-1 FALSE            61        97 section1                <span style="text-decoration: underline;">7</span>886</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 89 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This change maintains separation of sample_ids and is permitted</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_c.html" class="external-link">str_c</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="st">"_modified"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A SpatialExperiment-tibble abstraction: 99 × 7</span></span></span>
<span><span class="co">## <span style="color: #949494;"># </span><span style="color: #555555;">Features=6 | Cells=99 | Assays=counts</span></span></span>
<span><span class="co">##    .cell              in_tissue array_row array_col sample_id pxl_col_in_fullres</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> AAACAACGAATAGTTC-1 FALSE             0        16 section1…               <span style="text-decoration: underline;">2</span>312</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> AAACAAGTATCTCCCA-1 TRUE             50       102 section1…               <span style="text-decoration: underline;">8</span>230</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> AAACAATCTACTAGCA-1 TRUE              3        43 section1…               <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> AAACACCAATAACTGC-1 TRUE             59        19 section1…               <span style="text-decoration: underline;">2</span>519</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> AAACAGAGCGACTCCT-1 TRUE             14        94 section1…               <span style="text-decoration: underline;">7</span>679</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> AAACAGCTTTCAGAAG-1 FALSE            43         9 section1…               <span style="text-decoration: underline;">1</span>831</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> AAACAGGGTCTATATT-1 FALSE            47        13 section1…               <span style="text-decoration: underline;">2</span>106</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> AAACAGTGTTCCTGGG-1 FALSE            73        43 section1…               <span style="text-decoration: underline;">4</span>170</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAACATGGTGAGAGGA-1 FALSE            62         0 section1…               <span style="text-decoration: underline;">1</span>212</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> AAACATTTCCCGGATT-1 FALSE            61        97 section1…               <span style="text-decoration: underline;">7</span>886</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 89 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: pxl_row_in_fullres &lt;int&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This change does not maintain separation of sample_ids and produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span>sample_id <span class="op">=</span> <span class="st">"new_sample"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Error in .local(x, ..., value): Number of unique 'sample_id's is 2, but 1 was provided.</span></span></code></pre>
<p>The <code>pxl_col_in_fullres</code> and
<code>px_row_in_fullres</code> columns cannot be removed or modified
with <em>tidyverse</em> functions. This is consistent with the behaviour
of dimension reduction data in other <em>tidyomics</em> packages.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attempting to remove pxl_col_in_fullres produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="../reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">pxl_col_in_fullres</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `select_helper()`:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> Can't subset columns that don't exist.</span></span>
<span><span class="co">## <span style="color: #BB0000;">✖</span> Column `pxl_col_in_fullres` doesn't exist.</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attempting to modify pxl_col_in_fullres produces an error</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">pxl_col_in_fullres</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `dplyr::mutate()`:</span></span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> In argument: `pxl_col_in_fullres`.</span></span>
<span><span class="co">## <span style="font-weight: bold;">Caused by error:</span></span></span>
<span><span class="co">## <span style="color: #BBBB00;">!</span> object 'pxl_col_in_fullres' not found</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] tidySpatialExperiment_0.99.17   ggplot2_3.4.4                  </span></span>
<span><span class="co">##  [3] ttservice_0.4.0                 tidyr_1.3.0                    </span></span>
<span><span class="co">##  [5] dplyr_1.1.4                     tidySingleCellExperiment_1.12.0</span></span>
<span><span class="co">##  [7] SpatialExperiment_1.12.0        SingleCellExperiment_1.24.0    </span></span>
<span><span class="co">##  [9] SummarizedExperiment_1.32.0     Biobase_2.62.0                 </span></span>
<span><span class="co">## [11] GenomicRanges_1.54.1            GenomeInfoDb_1.38.5            </span></span>
<span><span class="co">## [13] IRanges_2.36.0                  S4Vectors_0.40.2               </span></span>
<span><span class="co">## [15] BiocGenerics_0.48.1             MatrixGenerics_1.14.0          </span></span>
<span><span class="co">## [17] matrixStats_1.2.0               BiocStyle_2.30.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] bitops_1.0-7              rlang_1.1.3              </span></span>
<span><span class="co">##  [3] magrittr_2.0.3            compiler_4.3.2           </span></span>
<span><span class="co">##  [5] DelayedMatrixStats_1.24.0 systemfonts_1.0.5        </span></span>
<span><span class="co">##  [7] vctrs_0.6.5               stringr_1.5.1            </span></span>
<span><span class="co">##  [9] pkgconfig_2.0.3           crayon_1.5.2             </span></span>
<span><span class="co">## [11] fastmap_1.1.1             magick_2.8.2             </span></span>
<span><span class="co">## [13] XVector_0.42.0            ellipsis_0.3.2           </span></span>
<span><span class="co">## [15] labeling_0.4.3            scuttle_1.12.0           </span></span>
<span><span class="co">## [17] utf8_1.2.4                rmarkdown_2.25           </span></span>
<span><span class="co">## [19] ragg_1.2.7                purrr_1.0.2              </span></span>
<span><span class="co">## [21] xfun_0.41                 zlibbioc_1.48.0          </span></span>
<span><span class="co">## [23] cachem_1.0.8              beachmat_2.18.0          </span></span>
<span><span class="co">## [25] jsonlite_1.8.8            highr_0.10               </span></span>
<span><span class="co">## [27] rhdf5filters_1.14.1       DelayedArray_0.28.0      </span></span>
<span><span class="co">## [29] Rhdf5lib_1.24.1           BiocParallel_1.36.0      </span></span>
<span><span class="co">## [31] parallel_4.3.2            R6_2.5.1                 </span></span>
<span><span class="co">## [33] bslib_0.6.1               stringi_1.8.3            </span></span>
<span><span class="co">## [35] limma_3.58.1              jquerylib_0.1.4          </span></span>
<span><span class="co">## [37] Rcpp_1.0.12               bookdown_0.37            </span></span>
<span><span class="co">## [39] knitr_1.45                R.utils_2.12.3           </span></span>
<span><span class="co">## [41] Matrix_1.6-1.1            tidyselect_1.2.0         </span></span>
<span><span class="co">## [43] abind_1.4-5               yaml_2.3.8               </span></span>
<span><span class="co">## [45] codetools_0.2-19          lattice_0.21-9           </span></span>
<span><span class="co">## [47] tibble_3.2.1              withr_3.0.0              </span></span>
<span><span class="co">## [49] evaluate_0.23             desc_1.4.3               </span></span>
<span><span class="co">## [51] pillar_1.9.0              BiocManager_1.30.22      </span></span>
<span><span class="co">## [53] plotly_4.10.4             generics_0.1.3           </span></span>
<span><span class="co">## [55] RCurl_1.98-1.14           sparseMatrixStats_1.14.0 </span></span>
<span><span class="co">## [57] munsell_0.5.0             scales_1.3.0             </span></span>
<span><span class="co">## [59] glue_1.7.0                lazyeval_0.2.2           </span></span>
<span><span class="co">## [61] tools_4.3.2               data.table_1.14.10       </span></span>
<span><span class="co">## [63] locfit_1.5-9.8            fs_1.6.3                 </span></span>
<span><span class="co">## [65] rhdf5_2.46.1              grid_4.3.2               </span></span>
<span><span class="co">## [67] DropletUtils_1.22.0       edgeR_4.0.9              </span></span>
<span><span class="co">## [69] colorspace_2.1-0          GenomeInfoDbData_1.2.11  </span></span>
<span><span class="co">## [71] HDF5Array_1.30.0          cli_3.6.2                </span></span>
<span><span class="co">## [73] textshaping_0.3.7         fansi_1.0.6              </span></span>
<span><span class="co">## [75] S4Arrays_1.2.0            viridisLite_0.4.2        </span></span>
<span><span class="co">## [77] gtable_0.3.4              R.methodsS3_1.8.2        </span></span>
<span><span class="co">## [79] sass_0.4.8                digest_0.6.34            </span></span>
<span><span class="co">## [81] SparseArray_1.2.3         dqrng_0.3.2              </span></span>
<span><span class="co">## [83] farver_2.1.1              rjson_0.2.21             </span></span>
<span><span class="co">## [85] htmlwidgets_1.6.4         memoise_2.0.1            </span></span>
<span><span class="co">## [87] htmltools_0.5.7           pkgdown_2.0.7            </span></span>
<span><span class="co">## [89] R.oo_1.25.0               lifecycle_1.0.4          </span></span>
<span><span class="co">## [91] httr_1.4.7                statmod_1.5.0</span></span></code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by William Hutchison, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
